csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs

```diff
@@ -0,0 +1,835 @@
+            current = current.ContainingType;
+        }
+
+        var baseCtor = @base.Constructors.
+            Where(c => !c.IsStatic).
+            Where(c =>
+                c.DeclaredAccessibility == Accessibility.Public ||
+                c.DeclaredAccessibility == Accessibility.Protected ||
+                c.DeclaredAccessibility == Accessibility.ProtectedOrInternal ||
+                containingTypes.Contains(c.ContainingType)
+            ).
+            OrderBy(c => c.Parameters.Length).FirstOrDefault();
+
+        return baseCtor?.Parameters.Length > 0 ? baseCtor : null;
+    }
+
+    private static IMethodSymbol? GetBaseConstructor(IMethodSymbol ctor)
+    {
+        if (ctor.MethodKind != MethodKind.Constructor)
+            return null;
+
+        return GetBaseConstructor(ctor.ContainingType);
+    }
+
+    private void StubParameters(ICollection<IParameterSymbol> parameters)
+    {
+        WriteCommaSep(parameters, parameter =>
+        {
+            switch (parameter.RefKind)
+            {
+                case RefKind.None:
+                    break;
+                case RefKind.Ref:
+                    stubWriter.Write("ref ");
+                    break;
+                case RefKind.Out:
+                    stubWriter.Write("out ");
+                    break;
+                case RefKind.In:
+                    stubWriter.Write("in ");
+                    break;
+                default:
+                    stubWriter.Write($"/* TODO: {parameter.RefKind} */");
+                    break;
+            }
+
+            if (parameter.IsParams)
+                stubWriter.Write("params ");
+
+            stubWriter.Write(parameter.Type.GetQualifiedName());
+            stubWriter.Write(" ");
+            stubWriter.Write(EscapeIdentifier(parameter.Name));
+
+            if (parameter.HasExplicitDefaultValue)
+            {
+                stubWriter.Write(" = ");
+                stubWriter.Write($"default({parameter.Type.GetQualifiedName()})");
+            }
+        });
+    }
+
+    private void StubMethod(IMethodSymbol symbol, IMethodSymbol? explicitInterfaceSymbol, IMethodSymbol? baseCtor)
+    {
+        var methodKind = explicitInterfaceSymbol is null ? symbol.MethodKind : explicitInterfaceSymbol.MethodKind;
+
+        var relevantMethods = new[] {
+                MethodKind.Constructor,
+                MethodKind.Conversion,
+                MethodKind.UserDefinedOperator,
+                MethodKind.Ordinary
+            };
+
+        if (!relevantMethods.Contains(methodKind))
+            return;
+
+        StubAttributes(symbol.GetAttributes());
+
+        StubModifiers(symbol, explicitInterfaceSymbol is not null);
+
+        if (IsUnsafe(symbol.ReturnType) || symbol.Parameters.Any(p => IsUnsafe(p.Type)))
+        {
+            stubWriter.Write("unsafe ");
+        }
+
+        if (explicitInterfaceSymbol is null && symbol.DeclaredAccessibility == Accessibility.Private)
```

Comment:
Maybe a comment to explain the logic behind this.

Reply:
Actually appears to be no longer needed.